### PR TITLE
Create gRPC servers and clients with TransportConnections.

### PR DESCRIPTION
**Pull request by nichite**
This allows callers to propvide their own transport configuration, such
as their own implementation of streams and sinks instead of sockets.

* Extract ConnectionServer from Server to support ServerTransportConnection
* Add ClientTransportConfig to support ClientTransportConnection
* Plumb ClientTransportConfig from Channel to Http2ClientConnection

### CODE REVIEW

**Comment by nichite** [comment]
Looks like some of the tests are timing out--I'm not sure why this isn't happening internally. Will need to look into it.

**Comment by mraleph** [comment]
This is due to this commit https://github.com/grpc/grpc-dart/commit/9ed03b6b969b9171d5b218e0ac308ae127c56984 which causes transport to endlessly attempt reconnecting on connection failures. This also causes https://github.com/grpc/grpc-dart/issues/339 - maybe we should just revert the commit that introduced reconnection logic for now.

File: lib/src/server/server.dart [hunk 1]
```diff
@@ -190,3 +160,49 @@ class Server {
     _secureServer = null;
   }
 }
+
+/// A gRPC server that serves via provided [ServerTransportConnection]s.
```
Comment (mraleph) [inline lib/src/server/server.dart:164]: I  find the ordering a bit confusing - I would expect base class to come ahead of child in the file

**Review by mraleph — APPROVED** [review]

**Comment by mhoad** [comment]
@nichite can I ask a dumb newbie question? Does this commit open the door to a non XHR based grpc-web future by opening up new possibilities like GRPC over Websockets and Streams or am I totally misreading this?

**Pull request merged**